### PR TITLE
fix(hooks): close format-staged config-sourcing and mixed-EOL follow-ups

### LINKED ISSUE #10500: Fall back to YAML parsing for extensionless .prettierrc configs
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

`verifyConfigDeps` only attempts `JSON.parse` for extensionless config files (`scripts/format-staged.mjs:265`), but Prettier accepts YAML in an extensionless `.prettierrc`. `strict` is true exactly when the staged config diverges from the worktree, so a diverged YAML `.prettierrc` now always throws `SkipError` and auto-staging is skipped for every formattable file — the diverged-config scenario the snapshot machinery from #10401 handled fine before this change (snapshot the raw blob, let Prettier's JSON-then-YAML loader parse it).

## Details

Fall back to the already-imported `parseYaml` for `ext === ''` — JSON first, then YAML, matching Prettier's own loader order — before declaring the config unparsable. Add a test: a diverged extensionless `.prettierrc` containing YAML should format (or at minimum not skip) rather than throw the misleading "cannot parse" notice.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236094 (claude)
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789228240 (Codex P2)

### LINKED ISSUE #10501: Verify relative dependencies for package.yaml configs
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

The `package.yaml` branch in `configSnapshotFor` (`scripts/format-staged.mjs:460`) returns `{ config: null, snapshotPath: null }` for a staged-clean file without ever verifying relative dependencies. Every other branch verifies: package.json runs `verifyDepSpecs` (line 454), JS configs run `verifyJsConfigDeps` (line 471), and the generic branch runs `verifyConfigDeps` (line 477). Prettier resolves a clean `package.yaml` — including relative `plugins` under its `prettier` key — from the working tree, so a plugin with unstaged edits formats the commit with uncommitted rules: exactly the provenance hole #10434 closes for every other config shape.

## Details

Parse the staged `package.yaml` blob, extract the `prettier` value Prettier actually reads, and run `verifyDepSpecs` on it (non-strict) before returning. Add a test: a staged-clean `package.yaml` whose relative plugin has unstaged edits should skip loudly instead of formatting.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236095 (claude)
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236096 (texra-ai)
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789228242 (Codex P2)

### LINKED ISSUE #10502: Verify JavaScript configs reached through package.json prettier pointers
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

When a package.json `prettier` string pointer resolves to a JavaScript target whose worktree copy equals its index blob, `configSnapshotFor` calls `verifyConfigDeps(targetPath, stagedTarget, false)` (`scripts/format-staged.mjs:450`). That function is a silent no-op for `.js`/`.cjs`/`.mjs` extensions in non-strict mode (lines 294–296), so the static import scan never runs. The direct JS path uses `verifyJsConfigDeps`; this pointer path does not, so a staged-clean `prettier.cjs` whose relative `import`/`require`/`plugins` entry has unstaged edits leaks those edits into the commit — the #10434 hole reached via the #10435 pointer path.

## Details

Dispatch on the target's extension before returning `{ config: targetPath }`: route JS targets through `verifyJsConfigDeps(targetPath, stagedTarget)`. Add a test with a clean pointer target plus an edited relative dependency, mirroring the existing #10434 tests.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236097 (claude)
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236098 (texra-ai)
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789228244 (Codex P2)

### LINKED ISSUE #10503: Handle package-name shareable configs in prettier string pointers
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

`configSnapshotFor` resolves every string `prettier` value with `resolve(dirname(worktreeConfigPath), stagedValue)` (`scripts/format-staged.mjs:426`). Prettier also accepts shareable-config package names here (e.g. `"prettier": "@company/prettier-config"`, resolved from node_modules — a documented configuration format). Those become `/repo/@company/prettier-config`, `readIndexFile` returns null, and the hook skips every affected file with a "not staged" notice even when package.json is clean and staged — a regression from the pre-#10401 behavior, where the snapshot preserved the package name and Prettier resolved it from node_modules.

## Details

Gate the pointer-following branch on `isRelativeSpecifier(stagedValue)` (or an in-repo absolute path). For package names, use the `configEqualsIndex` fast path, and skip loudly only if the worktree copy actually diverges. Add a test: a clean package.json with `"prettier": "@company/prettier-config"` (with a stub package resolvable in node_modules) should format rather than skip.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236099 (texra-ai)

### LINKED ISSUE #10504: Only verify config dependencies when a snapshot will be written
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

Relative-dependency verification runs unconditionally, before the `configEqualsIndex` gate, in all three branches that call it: package.json's object-value case (`scripts/format-staged.mjs:454`), the JS-config branch (line 471), and the generic branch (line 477). So it fires even when the config file itself is untouched and Prettier will load it straight from the working tree with no index snapshot involved. Concretely: an unmodified `.prettierrc.mjs` that imports `./opts.mjs` causes every commit touching any in-scope file to skip auto-staging whenever `opts.mjs` has unrelated unstaged edits — behavior that didn't exist before this PR. This also contradicts `verifyConfigDeps`' own doc comment ("when the config is clean we leave it to Prettier's own loader"); the `strict` flag only suppresses the unparsable-config skip, not `verifyConfigDep`'s divergent-dependency throw. The two new #10434 tests only cover the diverged-config case, so the regression isn't caught.

## Details

Decide the intended provenance policy and align the code with it. The suggested fix is to move the `if (configEqualsIndex) return { config: null, snapshotPath: null }` check before the `verify*Deps` call in each of the three branches, so dependency verification only runs when a snapshot is about to be written from index content. Note this pulls against the "verify more" direction of the package.yaml / pointer-target follow-ups — reconcile before implementing.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236100 (claude)

### LINKED ISSUE #10505: Write formatted output verbatim for fully staged mixed-EOL files
## Context

Follow-up from #10482 (fix(hooks): close format-staged config-sourcing follow-ups).

The fully-staged mixed-EOL branch in `mergeWorktree` (`scripts/format-staged.mjs:314-318`) writes `normalizeLf(formatted)` on the assumption that Prettier emits LF. With `endOfLine: "crlf"` — or `"auto"` over a CRLF-dominant staged blob — `formatted` contains CRLF, while the index was just staged with `formatted` verbatim. Normalizing the worktree write to LF then leaves the entire file as a new unstaged diff, defeating the #10432 goal of leaving no unstaged change for fully staged files.

## Details

Since the file is fully staged and the index holds `formatted` verbatim, write `formatted` verbatim in this branch (matching the newly staged blob) instead of normalizing it. Add a test: a fully staged mixed-EOL file formatted under `endOfLine: "crlf"` should leave `git diff` empty.

## References

- PR: #10482
- Review comment: https://github.com/LionSR/TeXRA/pull/10482#discussion_r3789236101 (texra-ai)

### PR DESCRIPTION
**Pull request by LionSR**
Closes #10500
Closes #10501
Closes #10502
Closes #10503
Closes #10504
Closes #10505

Six narrow follow-ups to `scripts/format-staged.mjs` from #10482, plus follow-up review fixes.

- #10500: extensionless `.prettierrc` parses with Prettier's YAML loader (JSON subset).
- #10501: package.yaml object configs verify relative dependencies; package.yaml string pointers mirror package.json pointer handling.
- #10502: JS configs reached through package.json/package.yaml pointers run the dependency scan with Node-accurate loader semantics.
- #10503: package-name shareable configs are left to Prettier when the manifest is clean, and skip loudly only when it diverges.
- #10504: relative-dependency verification only runs when a snapshot is needed; clean direct configs are left to Prettier's own loader.
- #10505: fully staged files write the formatted blob verbatim, except autocrlf checkouts.

Additional review fixes:
- CJS `require()` resolution matches Node order (exact, `.js`, `.json`, `.node`, then directory package main, then `index.js`/`.json`/`.node`); ESM imports and plugin paths stay exact.
- Directory package manifests must be tracked and index-equal before their `main` is trusted.
- Unstaged EOL-only conversions are preserved instead of being treated as fully staged.
- `core.autocrlf` is parsed as a Git boolean.

Verification (local, macOS arm64):

- `npx vitest run --config vitest.config.mjs src/test-kernel/scripts/FormatStaged.vitest.ts --testTimeout=60000 --hookTimeout=60000` → 60/60 passed
- `npm run typecheck` → passed
- `node --max-old-space-size=6144 ./node_modules/eslint/bin/eslint.js src` → passed
- `npx prettier --check scripts/format-staged.mjs src/test-kernel/scripts/FormatStaged.vitest.ts` → passed
- `npm run check:dead-code-ratchet` → passed (15 current = 15 baselined)
- `git diff --check` → clean

### CODE REVIEW

File: scripts/format-staged.mjs [hunk 1]
```diff
@@ -297,6 +323,17 @@ function verifyConfigDeps(configPath, configBlob, strict) {
   verifyDepSpecs(value, dirname(configPath));
 }
 
+/** Verify relative dependencies of a config reached through a package.json
+ * `prettier` string pointer. JavaScript targets need the JS import scan;
+ * other shapes go through the generic verifier. */
+function verifyConfigTargetDeps(configPath, configBlob) {
+  if (isJsConfig(configPath)) {
+    verifyJsConfigDeps(configPath, configBlob);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Implemented in `verifyConfigDep` so every relative dependency shape (JS imports, JSON/YAML plugins, package.yaml/package.json object deps) gets the same Node-style resolution before the index comparison: exact path first, then `.js`/`.cjs`/`.mjs`, then directory `package.json` `main`, then `index.js`/`.cjs`/`.mjs`; multiple candidates skip as ambiguous. Added regressions for extensionless `.js`, `.cjs`, directory package main, directory index, missing, ambiguous, and package.yaml. For `.cjs`/`.mjs` the hook no longer emits the false "not staged" skip; Prettier still reports its own loader error for Node-unresolvable specifiers, which is the correct separation.

File: scripts/format-staged.mjs [hunk 2]
```diff
@@ -464,18 +517,29 @@ function configSnapshotFor(worktreeConfigPath, stagedConfig, depth = 0) {
...
+    // before handing the working-tree copy back to Prettier. Parse errors are
+    // left to Prettier's own loader (this check is non-strict).
+    let stagedValue = null;
+    try {
+      stagedValue =
+        parseYaml(stripBom(stagedConfig.toString('utf8')))?.prettier ?? null;
+    } catch {
+      // ignored: Prettier will report the YAML error when it loads the file.
+    }
+    verifyDepSpecs(stagedValue, dirname(worktreeConfigPath));
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs:715]: Reconciled and documented in `configSnapshotFor`. The policy is now explicit: dependency verification runs only when the hook is about to hand Prettier a config path it derived from index content — a snapshot, or a pointer target resolved from a staged package.json/package.yaml. Clean configs returned as `null` are left to Prettier (`#10504`), with one deliberate exception: object-valued package.yaml can never be snapshotted, so its deps are verified before returning `null`; package.json/package.yaml string-pointer targets are verified because the hook resolves that chain itself. This keeps the six-issue batch consistent while explaining the distinction in-code.

File: scripts/format-staged.mjs [hunk 3]
```diff
@@ -297,6 +323,17 @@ function verifyConfigDeps(configPath, configBlob, strict) {
...
+/** Verify relative dependencies of a config reached through a package.json
+ * `prettier` string pointer. JavaScript targets need the JS import scan;
+ * other shapes go through the generic verifier. */
+function verifyConfigTargetDeps(configPath, configBlob) {
+  if (isJsConfig(configPath)) {
+    verifyJsConfigDeps(configPath, configBlob);
+    return;
+  }
+  verifyConfigDeps(configPath, configBlob, false);
+}
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Inlined. `verifyConfigTargetDeps` is gone; the single call site now dispatches directly between `verifyJsConfigDeps` and `verifyConfigDeps`.

File: scripts/format-staged.mjs [hunk 4]
```diff
@@ -254,36 +269,47 @@ function verifyJsConfigDeps(configPath, configBlob) {
...
-        throw new SkipError(
-          `cannot parse ${relToCwd(configPath)} to verify its relative ` +
-            'dependencies; skipped auto-staging.',
-        );
-      }
+      if (strict) unparsable();
       return;
     }
+  } else if (ext === '') {
+    // Prettier's extensionless loader accepts YAML as well as JSON; mirror
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Fixed: extensionless configs now parse with `parseYaml` directly, and the comment states Prettier uses its YAML loader (JSON is a subset).

File: scripts/format-staged.mjs [hunk 5]
```diff
@@ -464,18 +517,29 @@ function configSnapshotFor(worktreeConfigPath, stagedConfig, depth = 0) {
...
+      // ignored: Prettier will report the YAML error when it loads the file.
+    }
+    verifyDepSpecs(stagedValue, dirname(worktreeConfigPath));
     return { config: null, snapshotPath: null };
   }
 
   if (isJsConfig(worktreeConfigPath)) {
-    verifyJsConfigDeps(worktreeConfigPath, stagedConfig);
     if (configEqualsIndex) return { config: null, snapshotPath: null };
+    verifyJsConfigDeps(worktreeConfigPath, stagedConfig);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs:721]: This is the intended #10504 policy: clean configs returned as `null` are left to Prettier's working-tree loader, so a clean JS/package.json object/generic config with an unstaged dependency is not skipped. The policy is now documented in `configSnapshotFor`, and regression tests pin the hands-off behavior for clean JS, package.json object, and JSON/YAML configs.

File: scripts/format-staged.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs:715]: Implemented: the package.yaml branch now mirrors package.json string-pointer handling — it follows pointer paths, sources the target from the index (snapshotting when it diverges), and verifies the target's dependencies. A test pins the unstaged-target-dependency skip.

File: scripts/format-staged.mjs [hunk 6]
```diff
@@ -297,6 +323,17 @@ function verifyConfigDeps(configPath, configBlob, strict) {
   verifyDepSpecs(value, dirname(configPath));
 }
 
+/** Verify relative dependencies of a config reached through a package.json
+ * `prettier` string pointer. JavaScript targets need the JS import scan;
+ * other shapes go through the generic verifier. */
+function verifyConfigTargetDeps(configPath, configBlob) {
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Inlined at the call site; `verifyConfigTargetDeps` was removed.

File: scripts/format-staged.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs:715]: Documented as the deliberate exception in `configSnapshotFor`: object-valued package.yaml can never be snapshotted, so its dependencies are verified before returning `null`, while other clean configs follow the #10504 hands-off rule. The comment now states that rationale explicitly.

File: scripts/format-staged.mjs [hunk 7]
```diff
@@ -161,14 +180,98 @@ function isJsConfig(path) {
...
+
+/** Return the `main` entry of a dependency directory's package.json,
+ * preferring the staged index blob so an unstaged manifest cannot redirect
+ * extensionless resolution. Returns null when there is no usable manifest. */
+function readPackageMain(dir) {
+  const pkgPath = join(dir, 'package.json');
+  const pkgRel = relToCwd(pkgPath);
+  if (!pkgRel) return null;
+  let blob = readIndexFile(pkgRel);
+  if (!blob && existsSync(pkgPath)) blob = readFileSync(pkgPath);
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Fixed: `readPackageMain` now requires the directory package.json itself to be tracked and equal to its index blob before its `main` is used. Untracked manifests and unstaged manifest edits now skip loudly instead of redirecting resolution. Regressions cover both cases.

File: scripts/format-staged.mjs [hunk 8]
```diff
@@ -161,14 +180,98 @@ function isJsConfig(path) {
...
+  };
+
+  push(base);
+  for (const ext of JS_CONFIG_EXTENSIONS) push(`${base}${ext}`);
+
+  if (lstatSync(base, { throwIfNoEntry: false })?.isDirectory()) {
+    const main = readPackageMain(base);
+    if (main) push(resolve(base, normalizeSpecifier(main)));
+    for (const ext of JS_CONFIG_EXTENSIONS) {
+      push(join(base, `index${ext}`));
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Fixed: CJS directory resolution now checks package `main` first and only considers index.js/index.json/index.node when no usable main candidate resolves. Added a regression with both main.cjs and index.js present.

File: scripts/format-staged.mjs [hunk 9]
```diff
@@ -308,13 +417,16 @@ function mergeWorktree(path, stagedBlob, formatted) {
...
+  // is an autocrlf=true checkout, where Git checks the index's LF blob out as
+  // CRLF; preserve the tree's CRLF convention there.
   if (worktreeLf.equals(stagedLf)) {
-    const content = hasMixedEol(worktree)
-      ? normalizeLf(formatted)
-      : withWorktreeEol(formatted, worktree);
+    const content =
+      isAutocrlfCheckout() && worktree.includes(CRLF)
+        ? withWorktreeEol(formatted, worktree)
+        : formatted;
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Fixed: `mergeWorktree` now treats byte-different normalized equality as fully staged only when an autocrlf checkout explains it. Unstaged EOL-only conversions fall through to the merge path and are preserved; a regression pins LF index + CRLF worktree.

File: scripts/format-staged.mjs [hunk 10]
```diff
@@ -53,6 +61,8 @@ const NOTICE = '[format-staged]';
 
 const CRLF = Buffer.from('\r\n');
 
+const JS_CONFIG_EXTENSIONS = ['.js', '.cjs', '.mjs'];
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs]: Fixed: CommonJS LOAD_AS_FILE now includes .json (and .node), matching Node order. Added an extensionless require -> opts.json regression that formats successfully.

File: scripts/format-staged.mjs [hunk 11]
```diff
@@ -69,6 +79,15 @@ function git(args, { input } = {}) {
   return result.stdout;
 }
 
+/** True when `core.autocrlf=true`, so Git checks out the index's LF blobs
+ * as CRLF and a CRLF worktree convention must be preserved. */
+function isAutocrlfCheckout() {
+  const result = spawnSync('git', ['config', '--get', 'core.autocrlf'], {
+    encoding: 'utf8',
+  });
+  return result.status === 0 && result.stdout.trim() === 'true';
```
[the remark this answers is not in the document]
Comment (LionSR) [reply scripts/format-staged.mjs:91]: Fixed: `core.autocrlf` is now read with `git config --bool --get`, so 1/yes/on/True all canonicalize to true. Added a core.autocrlf=1 regression.
